GR Integration

Squigs44 opened the pull request:
This code attempts to integrate Ghost Ratings into webDiplomacy. Includes code that takes in game data and calculates GR adjustments. Integrates this code so that when games finish, GR is automatically adjusted. Provides an admin tool to backfill GR in batches. Has multiple table in the database to store GR historical data. This includes a game by game, user by user accounting of each GR adjustment (very detailed), as well as a month by month accounting of each user's GR (big picture). The config file has been adjusted to allow for customization of these GR calculations. You can create your own categories and filter by variant ID, press type, phase length, scoring type, etc.

As this is a large change, thorough testing and reviewing is needed. I have conducted a lot of testing on the actual math side of things and I am confident that the GR scoring is correct (more correct than the old scripts even), now it is mostly just making sure that this isn't breaking games and covers each of the unique scenarios correctly.

jmo1121109 commented on admin/adminActionsRestricted.php:
```diff
@@ -1152,6 +1158,42 @@ public function showApiKeys($params)
 		";
 	}
 
+	public function calculateGR($params) {
```
next line for { please

Squigs44 replied:
Almost made it through a whole PR without an indentation/style issue. One of these times I will get it.

Squigs44 replied:
Just put in a commit which addresses this.

jmo1121109 commented on admin/adminActionsRestricted.php:
```diff
@@ -1152,6 +1158,42 @@ public function showApiKeys($params)
+		{
+			return l_t('Invalid batch size. Please enter a positive integer.');
+		}
+		$data = $DB->sql_tabl("SELECT id, variantID, pressType, potType, turn, gameOver, phaseMinutes, processTime, playerTypes FROM wD_Games WHERE gameOver <> 'No' AND grCalculated <> 1 ORDER BY processTime ASC LIMIT ".$batch_size.";");
```
I know that the gameOver column and another column in this table have to be used carefully in queries because they aren't indexed. Lets get together at some point and test this select on the db for speed before pushing this.

Squigs44 replied:
Sounds good. Like most SQL related things, I defer to you completely.

jmo1121109 commented on ghostratings/calculations.php:
```diff
@@ -0,0 +1,346 @@
+     $this->k = 32;
+     $this->start = 100;
+     $this->pressMod = Config::$grPressMods[$this->pressType];
+     $this->modvalue = (17.5 * $this->variantMod * $this->pressMod);
```
What is 17.5? Can you throw in some comments around any of the number variables here llike 32, 100, etc. Will help anyone else like vDip who takes this to adjust a rating to their site needs.

Squigs44 replied:
All of these variables are just things used to control the scaling of the ranking systems: where a players rating starts at, how much they move up and down per game, etc. I will add comments on my next commit specifying this in the file.

Squigs44 replied:
Just put in a commit which addresses this.

jmo1121109 commented on gamemaster/game.php:
```diff
@@ -1050,6 +1051,33 @@ protected function setWon(Member $Winner)
 		global $DB;
 		$DB->sql_put("UPDATE wD_Games SET finishTime=".time()." WHERE id=".$this->id);
+		
+		//Do GR Stuff
```
Should all of the logic around GR be wrapped in a check that sees if the config section for GR exists? That way if a dev instance doesn't have it setup or another site chooses not to use it there's no issue?

Squigs44 replied:
Good idea, and shouldn't be a problem to implement. I'll add this in my next commit.

Squigs44 replied:
Just put in a commit which addresses this.

jmo1121109 commented on ghostratings/calculations.php:
```diff
@@ -0,0 +1,346 @@
+             }
+             else
+             {
+               $DB->sql_put("UPDATE wD_GhostRatingsHistory SET rating=".$newRating." WHERE userID=".$userID." AND categoryID=".$categoryID." AND monthYear=".$date);
```
add a limit 1 to the end of the update for safety. (or limit whatever the max number of updates should be)

Squigs44 replied:
It shouldn't be needed, but I will add it for safety.

Squigs44 replied:
Just put in a commit which addresses this.

jmo1121109 commented on ghostratings/calculations.php:
```diff
@@ -0,0 +1,346 @@
+         }
+       }
+     }
+     $DB->sql_put("UPDATE wD_Games SET grCalculated=1 WHERE id=".$this->gameID);
```
add limit 1 for safety.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on config.sample.php:
```diff
@@ -125,6 +125,79 @@ class Config
+			"variants" => array(1),
+			"presses" => array('Regular'),
+			"phases" => array('Nonlive'),
+			"scoring" => array('Winner-takes-all','Sum-of-squares')
```
Points Per Supply Center is included in each of the categories except for this one. This is to maintain the current categories. In the current scripts, PPSC is not included in this category while it is included in the other categories (I just double checked to make sure this is the case). The settings specified in the config match the settings in the current script.

jmo1121109 replied:
Right but I think the reason for that is the current script only calculates 1 month's worth. So when the only option on the site was PPSC, this category did include it. When it was WTA and PPSC only I also think it included it, it wasn't until PPSC was turned off that this category would have stopped. So to recreate this data via all games I think it has to be on.

Squigs44 replied:
The current script doesnt just run 1 month, thats why it is so slow. The way that GR is currently run, this specific category purposefully excludes all PPSC games in the history of webdip. If you want to change our categories to include PPSC, then we can do that, but if we are trying to maintain what the current GR scripts do, then this is how it is done. I can show you in the actual scripts if you would like.

jmo1121109 replied:
Sorry, I don't think I'm explaining this well. What I mean is that if you look at when ATC introduced SOS and removed PPSC in the code history it was in 2015 https://github.com/kestasjk/webDiplomacy/commits/5df717f1f7e166a5f417a58c37484d20a5cbe5cc/objects/scoringsystem.php

but if you go to the GR site https://sites.google.com/view/webdipinfo/ghost-ratings/ghost-ratings-by-category this category started in 2012. Back then it was calculated via some other criteria. What I'm trying to figure out is if it was just WTA or PPSC back then. If it was just WTA I think this will translate fine. But if it was PPSC then the historical data we calculate won't even be in the ballpark of what the current old GR monthly files show.

Squigs44 replied:
So I looked through some users in 2012 and which games count for this category. I confirmed that PPSC games were not counted towards this category in 2012. Only DSS games originally counted towards this category, and was later extended to include SOS when that was later introduced. So this should translate okay.

jmo1121109 replied:
Okay cool, thanks for digging into that

Squigs44 commented on config.sample.php:
```diff
@@ -125,6 +125,79 @@ class Config
+			"variants" => array(1,2,9,19,20,17),
+			/*pressMods sets which press type to include*/
+			"presses" => array('Regular','PublicPressOnly','NoPress','RulebookPress'),
+			/*phases sets whether you want to include live games, non-live games, or both. The cutoff is at 1 hour phase lengths*/
```
Any games with 1 hour phases are considered Non-Live, in order to match the current scripts. I will alter my comment here to specify that in my next commit with changes.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on config.sample.php:
```diff
@@ -125,6 +125,79 @@ class Config
+	 * An array of modvalues to use when calculating GhostRatings. The lower the number the more weight it carries
+	 * @var array
+	 */
+	public static $grVariantMods = array(1=>1, 2=>4, 9=>2, 19=>4, 20=>4, 15=>1, 23=>1, 17=>8);
```
This would be easy to modify, just let me know if you want them included or not. We would have to come up with a modifier for this, but using the existing world modifier of 8 seems reasonable for Known World..

Squigs44 commented on gamemaster/game.php:
```diff
@@ -1050,6 +1051,33 @@ protected function setWon(Member $Winner)
+		 $SCcounts[$Member->userID] = 0;
+		 $memberStatus[$Member->userID] = 'Defeated';
+	 }
+	 $memberStatus[$Winner->userID] = 'Won';
```
This whole section of code is getting the status of each player. It pulls their status mainly from the $this->Members->ByStatus, but it doesn't exactly match up with what we want. Since $this->Members->ByStatus gives us arrays of varying sizes, we loop through those. However, the $Winner variable here is not an array, it is a single user, so we don't need to loop. That line overwrites the previous logic that sets all remaining players as survived, singling out the winner from the other surviving players.

diptobi1 replied:
Is there a reason to not just pass the Members object to the GhostRatings instance? After the function calls Members->setWon() or Members->setDrawn() below, the member status and member SC count seem do be identical to ones you defined in your special GR arrays. So your "GR stuff" loops seem to be a bit redundant to me.

Squigs44 replied:
I wrote this about 8 or 9 months ago, so I don't remember all the details, but I wrote this like this to make backfilling GR much easier. When backfilling GR, creating game objects or member objects was difficult. I tried a few ways and even talked to JMO about it, but the way that those objects are set up, they are difficult to create and get you to do what you want outside of the established game engine (this is also why we added the variant info table in the database). Thus, by having an array of statuses and supply center counts, it adds a few loops inside of the setWon and setDrawn functions, but creates a way for me to go straight from the database when backfilling GR. Also, if we removed that part of the code, we would have to extract from the members array inside of the GR calculations file anyway. 

So while it is probably possible to cut out this section and save 17 lines of code by passing in the members array, it would add more convoluted code in other areas.

diptobi1 replied:
I understand.

My main concern are not the additional lines of code but having code at different places that does the same and is expected to stay the same. It could make further adjustments in the future more difficult as you need to remember to fix the (identical) code at any occurrence. Therefore I recommend using the existing status data of the Member objects instead of producing some new, i.e. directly use the status after the call Members->setDrawn()/setWon() (no matter how you pass it to the GR instance).

With the arrays needing to be provided in that way to the GR instance, let me suggest to use a different Members map instead:
After Members->setWon(), Members->setDrawn(), the following should in principle be equivalent to your code but uses the data calculated in the Members object and only one for loop:
`$SCcounts = array();`
`$memberStatus = array();`
`foreach($this->Members->ByOrder as $Member)`
`{`
`  $SCcounts[$Member->userID] = $Member->supplyCenterNo;`
`  $memberStatus[$Member->userID] = $Member->status;`
`}`
(not tested!)

This  7 lines would still be exactly identical setDrawn and setWon. For better maintainability you could consider to add an additional intermediate method that is called by both setWon and setDrawn to execute the identical GR code instead of having it in two places.

Squigs44 replied:
Okay, having a prepGR function that does what you described would be useful. I will switch things around to include this in my next commit.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on ghostratings/calculations.php:
```diff
@@ -0,0 +1,346 @@
+   private $pressMod;
+   private $modValue;
+   
+   private $k;
```
I will add a comment here in my next commit, but this controls how much a win/loss moves a player up or down specifically in the 1v1 rankings. It is commonly referred to as just 'k' in Elo Systems.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on install/1.65-1.66/update.sql:
```diff
@@ -0,0 +1,28 @@
+`categoryID` mediumint(8) unsigned NOT NULL,
+`monthYear` smallint(4) unsigned NOT NULL,
+`rating` FLOAT,
+INDEX ( `userID` )
```
I'm less familiar with adding indexes on SQL tables, so I will have to either Google this or ask for your help here.

jmo1121109 replied:
A good example here is the new tournament tables. Take a look at this one:

CREATE TABLE `wD_TournamentParticipants` (
`tournamentID` mediumint(8) unsigned NOT NULL,
`userID` mediumint(8) unsigned NOT NULL,
`status` enum('Applied','Accepted','Rejected', 'Left') NOT NULL,
INDEX ( `tournamentID` ),
INDEX ( `userID` )
) ENGINE=MyISAM DEFAULT CHARSET=utf8;

The important part with the index is that every column that will be used in a where clause of a  query in a large table should have an index. A single non indexed column will cause the query speed to decrease drastically.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on install/1.65-1.66/update.sql:
```diff
@@ -0,0 +1,28 @@
+  `timeFinished` int(10) unsigned NOT NULL
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
+
+ALTER TABLE `wD_Games` ADD COLUMN `grCalculated` INT NOT NULL DEFAULT 0;
```
Good catch.

Squigs44 replied:
Just put in a commit which addresses this.

Squigs44 commented on install/1.65-1.66/update.sql:
```diff
@@ -0,0 +1,28 @@
+`categoryID` mediumint(8) unsigned NOT NULL,
+`rating` FLOAT,
+`peakRating` FLOAT,
+`monthYear` smallint(4) unsigned NOT NULL,
```
Just put in a commit which addresses this.

Squigs44 commented:
Added the config boolean so that it can be set to auto calculate or not. Also added the index on grCalculated as requested. I tested the config boolean a few times and there were no problems. Check my sql though, since I didn't test that at all, but I think that is what you wanted? If not feel free to just change the sql to exactly what you want or let me know what I need to do to. Should have enough safeguards on this to be able to push it live without risk now.

Note that in order for the manual GR calculation mod tool to work, wD_VariantInfo has to be updated for each variant that is going to be included for GR. There is a mod tool that should automate this (although certain variants on vdip might break that mod tool), and I am pretty sure we have already ran that mod tool on all the webdip variants, I just wanted to put this down in writing here to double check, and in case someone else is trying to implement GR down the line and runs into this issue.
